src/django/api/models.py

```diff
@@ -2112,3 +2140,64 @@ def __str__(self):
         return (
             'Contributor ID: {contributor_id} - ' +
             'Column Name: {column_name}').format(**self.__dict__)
+
+
+@transaction.atomic
+def index_facilities(facility_ids=[]):
+    # If passed an empty array, create or update all existing facilities
+    if len(facility_ids) == 0:
+        print('Indexing all facilities...')
+        facility_ids = Facility.objects.all().values_list('id', flat=True)
+
+    contrib_type = 'facility_list_item__source__contributor__contrib_type'
+    contributor = 'facility_list_item__source__contributor'
+    list = 'facility_list_item__source__facility_list_id'
+    filter = (Q(is_active=True)
+              & Q(status__in=[FacilityMatch.AUTOMATIC,
+                              FacilityMatch.CONFIRMED,
+                              FacilityMatch.MERGED])
+              & Q(facility_list_item__source__is_active=True)
+              & Q(facility_list_item__source__is_public=True))
+
+    # Create a list of dictionaries in the structure of FacilityIndexes
+    data = FacilityMatch.objects \
+                        .filter(facility_id__in=facility_ids) \
+                        .annotate(name=F('facility__name'),
+                                  country_code=F('facility__country_code'),
+                                  location=F('facility__location')) \
+                        .values('name', 'country_code', 'location') \
+                        .annotate(contrib_types=ArrayAgg(
+                                    contrib_type,
+                                    filter=filter),
+                                  contributors=ArrayAgg(
+                                    contributor,
+                                    filter=filter),
+                                  lists=ArrayAgg(
+                                    list,
+                                    filter=filter),
+                                  id=F('facility_id'),
+                                  ppe=Concat('facility__ppe_product_types',
+                                             'facility__ppe_contact_phone',
+                                             'facility__ppe_contact_email',
+                                             'facility__ppe_website',
+                                             output_field=CharField()))
+
+    existing_indexes = FacilityIndex.objects \
+                                    .filter(id__in=facility_ids) \
+                                    .values_list('id', flat=True)
+
+    update_data = data.filter(id__in=existing_indexes)
+    update_fields = ['name', 'country_code', 'location', 'contrib_types',
+                     'contributors', 'ppe', 'lists']
+    FacilityIndex.objects.bulk_update(
+        [FacilityIndex(**kv) for kv in update_data], update_fields)
+
+    create_data = data.exclude(id__in=existing_indexes)
+    FacilityIndex.objects.bulk_create(
+        [FacilityIndex(**kv) for kv in create_data])
```

Comment:
I have about 11k facilities in my database. The first time I ran `./scripts/manage index_facilities` was about 8 seconds. Nearly all of the facilities in my database belong to the same contributor. When I updated the contributor record via the Django admin, the application locked up. I restarted tried running `./scripts/manage index_facilities` again and it ran for over 15 minutes before I forced  it to exit. With some `print` debugging I determined that it was getting stuck on the `bulk_update`. Switching to a delete followed by an insert resolved the issue and I think should have the same effect.

```suggestion
    FacilityIndex.objects.filter(id__in=facility_ids).delete()
    FacilityIndex.objects.bulk_create(
        [FacilityIndex(**kv) for kv in data])
```

Reply:
Great suggestion! It's both faster and shorter. I've updated to this in 0384630